twine/commands/upload.py

```diff
@@ -48,6 +50,47 @@ def skip_upload(response, skip_existing, package):
             (response.status_code == 403 and msg_403 in response.text)))
 
 
+def check_status_code(response, verbose):
+    """
+    Additional safety net to catch response code 410 in case the
+    UploadToDeprecatedPyPIDetected exception breaks.
+    Also includes a check for response code 405 and prints helpful error
+    message guiding users to the right repository endpoints.
+    """
+    if (response.status_code == 410 and
+            response.url.startswith(("https://pypi.python.org",
+                                     "https://testpypi.python.org"))):
+        raise exceptions.\
+            UploadToDeprecatedPyPIDetected(f"It appears you're uploading to "
+                                           f"pypi.python.org (or "
+                                           f"testpypi.python.org). You've "
+                                           f"received a 410 error response. "
+                                           f"Uploading to those sites is "
+                                           f"deprecated. The new sites are "
+                                           f"pypi.org and test.pypi.org. Try "
+                                           f"using {DEFAULT_REPOSITORY} (or "
+                                           f"{TEST_REPOSITORY}) to upload your"
+                                           f" packages instead. These are the "
+                                           f"default URLs for Twine now. More "
+                                           f"at https://packaging.python.org/"
+                                           f"guides/migrating-to-pypi-org/.")
+    elif response.status_code == 405 and "pypi.org" in response.url:
+        raise exceptions.PyPIMethodNotAllowed(f"You probably want one of these"
+                                              f"two URLs: {DEFAULT_REPOSITORY}"
+                                              f"or {TEST_REPOSITORY}. Check "
+                                              f"your --repository-url value.")
+    try:
+        response.raise_for_status()
+    except HTTPError as err:
```

Comment:
I'm still wondering if it's better to display the errors here, as I suggested in https://github.com/pypa/twine/pull/509#discussion_r333437457. Otherwise, I think seeing `You probably want one of these two URLs` *before* the `HTTP` error is a little confusing.

Alternatively, the message for the 405 error could include a description of the cause, similar to the 410 error. Something like `It appears you're trying to upload to PyPI, but you have an invalid URL.`.

Reply:
Since we're now raising an exception, we're not going to see `HTTPError`, right? We were seeing it before because we were printing the warnings/errors and then raising `HTTPError`.

```
(env) ➜  adityasaky-test-package twine upload --repository-url https://pypi.org/ dist/*
Enter your username: adityasaky
Enter your password: 
Uploading distributions to https://pypi.org/
Uploading adityasaky_test_package-0.0.1-py3-none-any.whl
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 17.3k/17.3k [00:00<00:00, 135kB/s]
InvalidPyPIUploadURL: You probably want one of these two URLs: https://upload.pypi.org/legacy/ or https://test.pypi.org/legacy/. Check your --repository-url value.
```

Reply:
Good call!

Reply:
> Since we're now raising an exception, we're not going to see `HTTPError`, right? We were seeing it before because we were printing the warnings/errors and then raising `HTTPError`.

Good call!